### PR TITLE
bytes2hexstring hook

### PR DESCRIPTION
**Pull request by RaoulSchaffranek**
This PR adds a new hook implementation for converting byte arrays into hex strings.
This is needed by the kontrol-node to encode memory as JSON strings efficiently.

### CODE REVIEW

File: runtime/strings/bytes.cpp [hunk 1]
```diff
@@ -162,6 +163,24 @@ SortBytes hook_BYTES_string2bytes(SortString s) {
   return hook_BYTES_bytes2string(s);
 }
 
+string *bytes2hexstring(string *b, size_t len) {
+  static const std::array<char, 16> hexchars
+      = {'0', '1', '2', '3', '4', '5', '6', '7',
+         '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
+  auto *result
+      = static_cast<string *>(kore_alloc_token(sizeof(string) + len * 2));
+  for (size_t i = 0; i < len; i++) {
```
Comment (tothtamas28) [inline runtime/strings/bytes.cpp]: I see that this pattern is repeated in `bytes.cpp`, but if `len > len(b)`, this loop will produce an out of bounds error. So consider either checking this as a precondition, or obtaining the length in the function body.

File: runtime/strings/bytes.cpp [hunk 2]
```diff
@@ -162,6 +163,24 @@ SortBytes hook_BYTES_string2bytes(SortString s) {
...
+    result->data[i * 2] = hexchars.at((b->data[i] >> 4) & 0xf);
+    result->data[i * 2 + 1] = hexchars.at(b->data[i] & 0xf);
+  }
+  init_with_len(result, len * 2);
+  return result;
+}
+
+SortString hook_BYTES_bytes2hexstring(SortBytes b) {
```
Comment (tothtamas28) [inline runtime/strings/bytes.cpp]: Consider `bytes2hex` for a snappier name.

Also, consider adding a doc comment.

File: unittests/runtime-strings/bytestest.cpp [hunk 3]
```diff
@@ -171,6 +172,19 @@ BOOST_AUTO_TEST_CASE(bytes2string) {
   BOOST_CHECK_EQUAL(0, memcmp(_1234->data, "1234", 4));
 }
 
+BOOST_AUTO_TEST_CASE(bytes2hexstring) {
+  auto empty = make_string("");
+  auto res = hook_BYTES_bytes2hexstring(empty);
+  BOOST_CHECK(res != empty);
+  BOOST_CHECK_EQUAL(len(res), 0);
+
+  auto _00ff = make_string("\x00\xff", 2);
```
Comment (tothtamas28) [inline unittests/runtime-strings/bytestest.cpp]: This is not a great test input, as it will mask the error where characters for a given byte are in the wrong order in the output. Maybe `"\x01\xef"`?

**Comment by tothtamas28** [comment]
@jberthold, is there a spreadsheet or similar for documenting hook implementations?

File: unittests/runtime-strings/bytestest.cpp (hunk 3 above)
Comment (tothtamas28) [reply unittests/runtime-strings/bytestest.cpp]: Also, consider adding a test case for other boundary values, in particular the empty string.

Comment (RaoulSchaffranek) [reply unittests/runtime-strings/bytestest.cpp]: Good catch! I improved the test case. The empty string is already covered.

File: runtime/strings/bytes.cpp [hunk 4]
```diff
@@ -1,4 +1,5 @@
 #include <algorithm>
+#include <array> // Add this include for std::array
```
Comment (tothtamas28) [inline runtime/strings/bytes.cpp]:
```suggestion
#include <array>
```

**Review by tothtamas28 — COMMENTED** [review]
LGTM. I'll leave the final approval to @Robertorosmaninho.

File: bin/llvm-krun [hunk 5]
```diff
@@ -9,6 +9,7 @@ expanded_input_file="$(mktemp tmp.in.XXXXXXXXXX)"
 parser_file="$(mktemp tmp.parse.XXXXXXXXXX)"
 temp_inputs=()
 
+# shellcheck disable=SC2329
```
Comment (Robertorosmaninho) [inline bin/llvm-krun:12]: Sorry, why is this needed here?

**Review by Robertorosmaninho — APPROVED** [review]
LGTM, just one question:

File: bin/llvm-krun (hunk 5 above)
Comment (RaoulSchaffranek) [reply bin/llvm-krun:12]: The shell check CI workflow kept failing, see for example this action: https://github.com/runtimeverification/llvm-backend/actions/runs/17319617915/job/49169763898

If I understand correctly, it's a false positive because the function is used in the following line. So I marked it to skip the check.

It's not related to this PR, I think, but maybe some CI flakiness caused by a shell check update.

Comment (Robertorosmaninho) [reply bin/llvm-krun:12]: Got it! Thanks for clarifying it!
